flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,263 @@
+        */
+      override def predictDataSet(
+        instance: KMeans,
+        predictParameters: ParameterMap,
+        testDS: DataSet[Vector])
+      : DataSet[LabeledVector] = {
+        instance.centroids match {
+          case Some(centroids) =>
+            testDS.mapWithBcVariable(centroids) {
+              (dataPoint, centroids) => selectNearestCentroid(dataPoint, centroids)
+            }
+          case None =>
+            throw new RuntimeException("The KMeans model has not been trained. Call first fit" +
+              "before calling the predict operation.")
+        }
+      }
+    }
+  }
+
+  /**
+    * [[FitOperation]] which iteratively computes centroids that match the given input DataSet by
+    * adjusting the given initial centroids.
+    *
+    * @return A new  [[FitOperation]] to train the model using the training data set.
+    */
+  implicit def fitKMeans = {
+    new FitOperation[KMeans, Vector] {
+      override def fit(
+        instance: KMeans,
+        fitParameters: ParameterMap,
+        trainingDS: DataSet[Vector])
+      : Unit = {
+        val resultingParameters = instance.parameters ++ fitParameters
+
+        val centroids: DataSet[Seq[LabeledVector]] = trainingDS.
+          getExecutionEnvironment.fromElements(resultingParameters.get(InitialCentroids).get)
+        val numIterations: Int = resultingParameters.get(NumIterations).get
+
+        val finalCentroids = centroids.iterate(numIterations) {
+          currentCentroids =>
+            val newCentroids: DataSet[LabeledVector] = trainingDS
+              .mapWithBcVariable(currentCentroids) {
+                (dataPoint, centroids) => selectNearestCentroid(dataPoint, centroids)
+              }
+              .map(x => (x.label, x.vector, 1.0)).withForwardedFields("label->_1; vector->_2")
+              .groupBy(x => x._1)
+              .reduce(
+                (p1, p2) => (p1._1, (p1._2.asBreeze + p2._2.asBreeze).fromBreeze, p1._3 + p2._3)
+              )
+              .withForwardedFields("_1")
+              .map(
+                x => {
+                  BLAS.scal(1.0 / x._3, x._2)
+                  LabeledVector(x._1, x._2)
+                })
+              .withForwardedFields("_1->label")
+
+          // currentCentroids contains only one element. So, this is output only once
+          currentCentroids.mapWithBcSet(newCentroids) {
+            (_, newCenters) => newCenters
+          }
+        }
+
+        instance.centroids = Some(finalCentroids)
+      }
+    }
+  }
+
+  /**
+    * Converts a given vector into a labeled vector where the label denotes the label of the closest
+    * centroid.
+    *
+    * @param dataPoint The vector to determine the nearest centroid.
+    * @param centroids A collection of the centroids.
+    * @return A [[LabeledVector]] consisting of the input vector and the label of the closest
+    *         centroid.
+    */
+  @ForwardedFields(Array("*->vector"))
```

Comment:
I am however not very familiar with the use of semantic annotations. Someone should verify this explicitly.